File: TsinghuaNet.Helper/Helpers/EnumerableEx.vb
Imports System.Runtime.CompilerServices

Public Module EnumerableEx
    <Extension>
    Public Function OrderBy(Of T, TKey)(source As IEnumerable(Of T), selector As Func(Of T, TKey), descending As Boolean) As IOrderedEnumerable(Of T)
        If descending Then
            Return source.OrderByDescending(selector)
        Else
            Return source.OrderBy(selector)
        End If
    End Function

    <Extension>
    Public Function Sum(Of T)(source As IEnumerable(Of T), selector As Func(Of T, ByteSize)) As ByteSize
        Return New ByteSize(source.Sum(Function(v) selector(v).Bytes))
    End Function
End Module

File: TsinghuaNet/Helper/EnumerableEx.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace TsinghuaNet.Helper
{
    public static class EnumerableEx
    {
        public static IOrderedEnumerable<T> OrderBy<T, TKey>(this IEnumerable<T> source, Func<T, TKey> selector, bool descending)
            => descending ? source.OrderByDescending(selector) : source.OrderBy(selector);

        public static ByteSize Sum<T>(this IEnumerable<T> source, Func<T, ByteSize> selector)
            => new ByteSize(source.Sum(v => selector(v).Bytes));
    }
}
